### PR TITLE
Preemption e2e test

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

A follow up on #70898 to ensure code coverage.

**Special notes for your reviewer**:

Integration test is really challenging to reproduce original issue because:

- Preemptor pod is (always) head of activeQ, and will be firstly re-tried

	This is because when preemptor pod is trying to preempt low-priority pods, it's firstly updated (status.NominatedNode), hence put back into activeQ (a max-heap) at the head.

- Retry of preemptor pod is always succeeded

	This is because in integration, kubelet isn't involved, you have to set deleteGracePeriod=0 to get the pod deleted (see [slack link](https://kubernetes.slack.com/archives/C0EG7JC6T/p1542583956347100) for details). But with that, it's actually not mimicing the real world - in real world, pod have terminationGracePeriod, so they get into Terminating status first, and then be deleted. Say they're sort of deleted asynchronously. But in integration test, it's sort of synchronous (required to have 0 terminationGracePeriod)

Based on above reasons, I switched back to using an e2e test to simulate the issue. And I've tweaked it from checking podEvents to checking number of "podNamesSeen" of innocent pods only, which should be good.

BTW, there is a separate commit to address https://github.com/kubernetes/kubernetes/pull/70898#discussion_r233679741.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/sig scheduling

### CODE REVIEW

**Comment by Huang-Wei** [comment]
I suppose the plan is to merge it in 1.13, so

/priority critical-urgent

**Comment by k82cn** [comment]
still for 1.13?

**Comment by Huang-Wei** [comment]
Friendly ping @bsalamat..

File: test/e2e/scheduling/preemption.go [hunk 1]
```diff
@@ -353,3 +371,224 @@ var _ = SIGDescribe("PodPriorityResolution [Serial]", func() {
...
+		runPauseRS(f, rsConfs[rsNum-1])
+
+		framework.Logf("pods created so far: %v", podNamesSeen)
+		framework.Logf("length of pods created so far: %v", len(podNamesSeen))
+
+		// count pods number of RepliaSet3, if it's more than orignal replicas (4)
+		// then means its pods has been preempted once or more
+		rs3PodsSeen := 0
+		for podName := range podNamesSeen {
+			if strings.HasPrefix(podName, "rs-pod3") {
```
Comment (bsalamat) [inline test/e2e/scheduling/preemption.go]: Why does this apply only to RS3? It should apply to RS1 and RS2 as well, right? We should ensure that no more than 10 replicas of RS1 and no more than 8 replicas of RS2 exist and all those replicas are either deleted or unscheduled.

File: test/e2e/scheduling/preemption.go [hunk 2]
```diff
@@ -353,3 +371,224 @@ var _ = SIGDescribe("PodPriorityResolution [Serial]", func() {
...
+			}
+		}
+		if rs3PodsSeen != 4 {
+			framework.Failf("some pods of ReplicaSet3 have been preempted: expect 4 pod names, but got %d", rs3PodsSeen)
+		}
+	})
+
+})
+
+type pauseRSConfig struct {
```
Comment (bsalamat) [inline test/e2e/scheduling/preemption.go]: We may want to move everything below this line to util.go.

**Review by bsalamat — COMMENTED** [review]
Thanks, @Huang-Wei!

File: test/e2e/scheduling/preemption.go [hunk 3]
```diff
@@ -353,3 +371,224 @@ var _ = SIGDescribe("PodPriorityResolution [Serial]", func() {
...
+		node, err = cs.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+		if err != nil {
+			framework.Failf("error getting node %q: %v", nodeName, err)
+		}
+
+		// update Node API object with a fake resource
+		nodeCopy := node.DeepCopy()
+		// force it to update
+		nodeCopy.ResourceVersion = "0"
+		nodeCopy.Status.Capacity[fakecpu] = resource.MustParse("800")
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply test/e2e/scheduling/preemption.go]: It stays. Kubelet seems to have no interests on this kind of extended resource.

File: test/e2e/scheduling/preemption.go [hunk 4]
```diff
@@ -353,3 +371,224 @@ var _ = SIGDescribe("PodPriorityResolution [Serial]", func() {
...
+		)
+		go podController.Run(stopCh)
+		defer close(stopCh)
+
+		// prepare four ReplicaSet
+		rsConfs := []pauseRSConfig{
+			{
+				Replicas: int32(5),
+				PodConfig: pausePodConfig{
+					Name:              fmt.Sprintf("pod1"),
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply test/e2e/scheduling/preemption.go]: lol, indeed..

File: test/e2e/scheduling/preemption.go (hunk 1 above)
Comment (Huang-Wei) [reply test/e2e/scheduling/preemption.go]: True, that's exactly what I did in previous commit... (thought it's subtle)

Will update.

File: test/e2e/scheduling/preemption.go [hunk 5]
```diff
@@ -353,3 +371,224 @@ var _ = SIGDescribe("PodPriorityResolution [Serial]", func() {
...
+		// create ReplicaSet4
+		// if runPauseRS failed, it means ReplicaSet4 cannot be scheduled even after 1 minute
+		// which is unacceptable
+		runPauseRS(f, rsConfs[rsNum-1])
+
+		framework.Logf("pods created so far: %v", podNamesSeen)
+		framework.Logf("length of pods created so far: %v", len(podNamesSeen))
+
+		// count pods number of RepliaSet3, if it's more than orignal replicas (4)
+		// then means its pods has been preempted once or more
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply test/e2e/scheduling/preemption.go]: I don't see how GC is involved. Isn't this just (1) scheduler preempts (call `clientset.Delete()` to delete) pod(s), and then (2) RelicaSet controller tries to reconcile to spawn new pod(s)?

File: test/e2e/scheduling/preemption.go (hunk 2 above)
Comment (Huang-Wei) [reply test/e2e/scheduling/preemption.go]: which `util.go` do you mean specifically?

Comment (bsalamat) [reply test/e2e/scheduling/preemption.go]: Never mind. We have a util.go in our integration tests, but not in e2e tests.

**Review by bsalamat — COMMENTED** [review]
/lgtm
/approve

Thanks, @Huang-Wei!

**Pull request merged**